src/RequirementChecker/AppRequirementsFactory.php

```diff
@@ -0,0 +1,204 @@
+                        $extension
+                    );
+                }
+
+                $requirements->addRequirement(
+                    self::generateExtensionCheckStatement($extension),
+                    $message,
+                    $helpMessage
+                );
+            }
+        }
+    }
+
+    private static function exportRequirementsIntoConfig(RequirementCollection $requirements): array
+    {
+        return array_map(
+            function (Requirement $requirement): array {
+                return [
+                    $requirement->getIsFullfilledChecker(),
+                    $requirement->getTestMessage(),
+                    $requirement->getHelpText(),
+                ];
+            },
+            $requirements->getRequirements()
+        );
+    }
+
+    /**
+     * Collects the extension required. It also accounts for the polyfills, i.e. if the polyfill `symfony/polyfill-mbstring` is provided
+     * then the extension `ext-mbstring` will not be required.
+     *
+     * @param array $composerLockContents
+     *
+     * @return array Associative array containing the list of extensions required
+     */
+    private static function collectExtensionRequirements(array $composerLockContents): array
+    {
+        $requirements = [];
+        $polyfills = [];
+
+        $platform = $composerLockContents['platform'] ?? [];
+
+        foreach ($platform as $package => $constraint) {
+            if (preg_match('/^ext-(?<extension>.+)$/', $package, $matches)) {
+                $extension = $matches['extension'];
+
+                $requirements[$extension] = [self::SELF_PACKAGE];
+            }
+        }
+
+        $packages = $composerLockContents['packages'] ?? [];
+
+        foreach ($packages as $packageInfo) {
+            $packageRequire = $packageInfo['require'] ?? [];
+
+            if (1 === preg_match('/symfony\/polyfill-(?<extension>.+)/', $packageInfo['name'], $matches)) {
+                $extension = $matches['extension'];
+
+                if ('php' !== substr($extension, 0, 3)) {
+                    $polyfills[$extension] = true;
+                }
+            }
+
+            foreach ($packageRequire as $package => $constraint) {
+                if (1 === preg_match('/^ext-(?<extension>.+)$/', $package, $matches)) {
+                    $extension = $matches['extension'];
+
+                    if (false === array_key_exists($extension, $requirements)) {
+                        $requirements[$extension] = [];
+                    }
+
+                    $requirements[$extension][] = $packageInfo['name'];
+                }
+            }
+        }
+
+        return array_diff_key($requirements, $polyfills);
+    }
+
+    private static function generatePhpCheckStatement(string $requiredPhpVersion): string
+    {
+        return "return require_once 'Semver.php'; \\_HumbugBox\Composer\Semver::satisfies(preg_replace('#^([^~+-]+).*$#', '$1', \PHP_VERSION), '$requiredPhpVersion');";
```

Comment:
After some research, as our minimum PHP version is 5.3, this part can be simplify to:

```php
return "return require_once 'Semver.php'; \\_HumbugBox\Composer\Semver::satisfies(sprintf('%d.%d.%d', \PHP_MAJOR_VERSION, \PHP_MINOR_VERSION, \PHP_RELEASE_VERSION), '$requiredPhpVersion');";
```
No need of a RegExp to parse the version. Those constants have been [added in PHP 5.2.7](https://secure.php.net/manual/en/function.phpversion.php#refsect1-function.phpversion-examples).

-----

Tested on:
 - Ubuntu:
```php
echo PHP_VERSION; // output: 7.0.28-0ubuntu0.16.04.1
echo sprintf('%d.%d.%d', \PHP_MAJOR_VERSION, \PHP_MINOR_VERSION, \PHP_RELEASE_VERSION); // output: 7.0.28
```
 - Docker (dockette/php71):
```php
echo PHP_VERSION; // output: 7.1.15-1+0~20180306120016.15+stretch~1.gbp78327e
echo sprintf('%d.%d.%d', \PHP_MAJOR_VERSION, \PHP_MINOR_VERSION, \PHP_RELEASE_VERSION); // output: 7.1.15

```